### ISSUE TITLE
Use merge squash for a clean history

### ISSUE THREAD

**Issue by ylecuyer** [FredrikNoren/ungit#1319]
Hello,

It is hard to follow the changes only watching at the master branch, could @maintainers start enforcing merge squash to have a clean history ?

how it looks now :

![Screenshot_2020-04-24_09-27-42](https://user-images.githubusercontent.com/1866809/80186191-e43b2680-860d-11ea-8685-ef2dd2e5195b.png)

**Comment by wmertens** [comment]
I much prefer having the history of the merged branch available, so you can see the logical steps the change does, and later you can blame some code and see why it was changed.

This of course means that merged branches should have nice commits, squashing WiP commits and separating concerns of course.

To get the effect you want, is it not enough to change the merge message so it explains the merged branch?

**Comment by Hirse** [comment]
Afaik, GitHub (now) also supports rebase merges, which would keep the individual commits, but not produce a separate merge commit. This might make sense depending on the feature branch.

![image](https://user-images.githubusercontent.com/2564094/80297526-a1c23880-8738-11ea-82ae-a14bbae4f5c9.png)

**Comment by wmertens** [comment]
@Hirse but then you have a bunch of bundled commits that you unbundle, and you lose some information about the grouping

**Comment by jung-kim** [comment]
Honestly, I've never saw a value of merging many commits into a single giant commits, which is what squash does.  Because if you want to see single consolidated diff, one can do that with git diff and I'd much rather see each git commits as they can be helpful in reviews and revisit why each, small, changes were made.

**Comment by wmertens** [comment]
🤔 actually, isn't the main problem that merge commits automatically get an informationless message?

How about, when merging, add all the commit subjects as a markdown list into the merge commit message editor so they can be summarized or left as-is?

And/or, in the UI, show this automatically for commits with multiple parents? When you hover over the merge commit, it shows the subjects on the side branch too?

**Comment by ylecuyer** [comment]
> Honestly, I've never saw a value of merging many commits into a single giant commits, which is what squash does. Because if you want to see single consolidated diff, one can do that with git diff and I'd much rather see each git commits as they can be helpful in reviews and revisit why each, small, changes were made.

It is the oposite for me, when working on big projects with a team, squashing merges has been the most helpful thing I miss in other projects.

If you blame some line of code you can directly see the commit and the whole unit of work which led to this change.
Plus, if the commit message is properly set, you can find all the related discussion on the PR or issue linked to this commit.

On the contrary when reviewing the history of a project it has never been usefull to me to see a lot of commits like "trying something", "I hope the CI is green now".

Regarding @wmertens proposal of having semi-squashed commits in the history, this is often a clear indicator that the PR is too big and must be split in chunks (i.e. multiple PRs with steps)

Btw it looks like the latest PRs have been merge squashed and you can see by yourselves that the quality of the git log has dramaticaly improved since a71b81f2 : https://github.com/FredrikNoren/ungit/commits/master

**Comment by jung-kim** [comment]
If you feel that way, there is nothing I can do about that.  You do realize if you want to see git diff between a hash and master, no matter how many commit a hash is ahead of, you can do so with git diff right?  i.e. `https://github.com/FredrikNoren/ungit/compare/e595e0c..master` that will show diff between hash `e595e0c` and current master.

Again, if you want to see squashed diffs, you can do that.  But once commits are squashed you cannot really break them out.  What is the value of enforcing squash?

**Comment by wmertens** [comment]
Point in case: @campersau merged the prettier PR (yey!) but squashed it. The result is a single commit where you can't tell the difference between automated formatting and actual changes needed to configure prettier:

https://github.com/FredrikNoren/ungit/commit/ac32c679ed398879747adcf8f5a8caf5729fcad2

IMHO it would have been better to merge normally.

**Comment by ylecuyer** [comment]
Indeed, this kind of enormous automated change to a whole repository is not something recommended. 

The approach taken by pronto (https://github.com/prontolabs/pronto / I don't know if there is something similar in js) is better as it is incremental. The code gets formatted/cleaned as it is edited and more important it doesn't kill the blame functionality (now if we blame a file we will for the majority of cases find the prettier commit)

**Comment by wmertens** [comment]
@ylecuyer well, that's not that much of a problem, you just look for the blame before that. If you do it incrementally then there will be a lot of commits with irrelevant changes. Better to do it all in one go, and any pending PRs can do the same change easily.
